libs/core/type_support/include/hpx/type_support/is_trivially_relocatable.hpp

```diff
@@ -0,0 +1,37 @@
+//  Copyright (c) 2023 Isidoros Tsaousis-Seiras
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <type_traits>
+
+// Macro to specialize template for given type
+#define HPX_DECLARE_TRIVIALLY_RELOCATABLE(T)                                   \
+    namespace hpx {                                                            \
+        template <>                                                            \
+        struct is_trivially_relocatable<T> : std::true_type                    \
+        {                                                                      \
+        };                                                                     \
+    }
+
+namespace hpx {
+
+    template <typename T, typename = void>
+    struct is_trivially_relocatable : std::false_type
+    {
+    };
+
+    // All trivially copyable types are trivially relocatable
+    template <typename T>
+    struct is_trivially_relocatable<T,
+        std::enable_if_t<std::is_trivially_copyable_v<T>>> : std::true_type
```

Comment:
Do we have to disable this specialization if `T` was marked as trivially relocatable (by specializing `is_trivially_relocatable<T>` where `T` is also trivially copyable? 

Alterntively, we could change the macro above to:
```
#define HPX_DECLARE_TRIVIALLY_RELOCATABLE(T)                                   \
    namespace hpx {                                                            \
        template <>                                                            \
        struct is_trivially_relocatable<T,                                     \
            std::enable_if_t<!std::is_trivially_copyable_v<T>>>                \
          : std::true_type                                                     \
        {                                                                      \
        };                                                                     \
    }
```

Reply:
Alternatively, just change the base case to
```
    template <typename T, typename = void>
    struct is_trivially_relocatable : std::is_trivially_copyable<T>
    {
    };
```

Reply:
Thank you that's very elegant, I will apply this. Would 

std::enable_if_t<!std::**is_trivially_relocatable_v**<T>>> ...

work as a more general solution? It does look kind of suspicious...

Reply:
No, you don't want to define `is_trivially_relocatable` in terms of itself. ;)
Notice that P1144 can't help with this part of the HPX design, because P1144 proposes that the core language should just know which types are trivially relocatable — P1144's raison d'etre is specifically to _avoid_ this kind of opt-in macro. (But HPX needs this part _because_ P1144's core-language changes aren't in the language yet.)

@isidorostsa: You should definitely take a look at what BSL and Folly are doing with their opt-in macros, and try to evaluate what advantages and disadvantages each approach might have. I think what you're doing with just letting the user specialize `hpx::is_trivially_relocatable<T, enable_if_t<...>>` is fine, but I don't think it is isomorphic to either BSL's or Folly's approach, so it would be good to evaluate, justify, and document.

* [`bslmf::IsBitwiseMoveable`](https://bloomberg.github.io/bde-resources/doxygen/bde_api_prod/bslmf__isbitwisemoveable_8h_source.html)
* [`folly::IsRelocatable`](https://github.com/facebook/folly/blob/5c227af84c1379f6f196932ca6b2fcfc7863c8ca/folly/Traits.h#L535)

Reply:
Thanks for this note, having examined `bslmf` and `folly` we decided to avoid using a nested trait, due to it being invasive to existing code to work. We did implement more generic macros to compensate for the lack of template and conditional support by the previous macro.

new macros: https://github.com/STEllAR-GROUP/hpx/pull/6264/commits/7139cb18df90897ecf39a1345c30faeb950b9751